src/main/java/org/simplify4u/plugins/keyserver/PGPKeysCache.java

```diff
@@ -175,7 +178,46 @@ public PGPPublicKeyRing getKeyRing(KeyId keyID) throws IOException {
             }
 
             // key not exists in cache or something wrong with cache, so receive from servers
-            return keyServerList.execute(keysServerClient -> receiveKey(keyFile, keyID, keysServerClient));
+            return Try.of(() -> keyServerList.execute(keysServerClient -> receiveKey(keyFile, keyID, keysServerClient)))
+                    .onFailure(PGPKeyNotFound.class, e -> writeNotFoundCache(path))
+                    .get();
+        }
+    }
+
+    private void writeNotFoundCache(String keyFilePath) {
+
+        File file = new File(cachePath, keyFilePath + ".404");
+
+        try {
+            Files.write(file.toPath(), String.valueOf(System.currentTimeMillis()).getBytes(US_ASCII));
+        } catch (IOException e) {
+            LOGGER.warn("Write file: {} exception: {}", file, getMessage(e));
+            deleteFile(file);
+        }
+    }
+
+    private void checkNotFoundCache(String keyFilePath) throws PGPKeyNotFound {
+
+        File file = new File(cachePath, keyFilePath + ".404");
+
+        if (file.isFile()) {
+
+            long markTime = Try.of(() -> {
+                        byte[] cacheContent = Files.readAllBytes(file.toPath());
+                        return Long.parseLong(new String(cacheContent, US_ASCII));
+                    })
+                    .onFailure(e -> LOGGER.warn("Read cache file: {}", file, e))
+                    .getOrElse(0L);
+
+            long elapsedTimeH = (System.currentTimeMillis() - markTime) / 60 * 60 * 1000;
```

Comment:
```suggestion
            long elapsedTimeH = (System.currentTimeMillis() - markTime) / 60 / 60 / 1000;
```

Reply:
fixed .. I used `Duration` instead of manually calculated.
Thanks.